### ISSUE TITLE
Memory leak issue

### ISSUE THREAD

**Issue by TuringTW** [tudo-astroparticlephysics/PROPOSAL#109]
### In UtilityDecorator
**Description**
I found a memory leak problem when I try to test the ionization and photonuclear process of tau. After my inspection via valgrind, I think the leakage is caused by the UtilityDecorator. Under this special case, the program will frequently trigger the function at PropagationUtilityInterpolant.cxx:[604](https://github.com/tudo-astroparticlephysics/PROPOSAL/blob/master/private/PROPOSAL/propagation_utility/PropagationUtilityInterpolant.cxx#L604):

`return UtilityIntegralContRand(utility_).Calculate(ei, ef, rnd);`

which will generate a temporal `UtilityIntegralContRand`. However, the base class of it, which is `UtilityDecorator`, has a memory leak. The `utility_` variable in the `UtilityDecorator` will be clone in the constructor of `UtilityDecorator`, however, it won't be released during the deconstruction of `UtilityDecorator` because it stores as a reference variable. 

**How to reproduce**
[Here](https://gist.github.com/TuringTW/ff5c38ee15efbd9accb75f2a7645d07a)'s the config file and the sample code I used.

**Workaround**
The problem can be solved by adding a `delete &utility_` in the deconstructor of `UtilityDecorator`. 
```
UtilityDecorator::~UtilityDecorator() {
    delete &utility_;
}
```
But deleting a reference through its address is suggested according to C++ standard. I think of changing it from reference to normal variable, that is 

from `const Utility& utility_;` to `const Utility utility_;`

However, this will introduce one more copy process each time during the construction of `UtilityDecorator`, which may slow down the program a little bit. I think it may be better to store the `utility_` as a (smart) pointer. But it require a huge modification of the code so I didn't do it.

### In Medium.cxx
**Description**
During my inspection of the problem above, I found a minor memory leak in the Medium.cxx:341
```
void Medium::SetDensityDistribution(Density_distr& density_distr) {
    dens_distr_ = density_distr.clone();
}
```
After setting the new density distribution, the original one won't be released. Therefore, I add a line of code to delete the original one to solve this leakage.
```
void Medium::SetDensityDistribution(Density_distr& density_distr) {
    delete distr_;
    dens_distr_ = density_distr.clone();
}
```

### In XXXXXXXFactory.cxx
**Description**
Take AnnihilationFactory for example, when the program find a corresponding enum of the parameterization, 
```
return new AnnihilationIntegral(*it->second(particle_def, medium, def.multiplier));
```
The code `it->second(particle_def, medium, def.multiplier)` will generate a pointer of annihilation parameter. After passing the parameter to create a `AnnihilationIntegral`, this pointer is not released. I think this problem can be solved by using a smart pointer
```
std::unique_ptr<Annihilation> param(it->second(particle_def, medium, def.multiplier)); 
return new AnnihilationIntegral(*param);
```
### Modification of code related to the issues above
[here](https://github.com/tudo-astroparticlephysics/PROPOSAL/compare/master...TuringTW:memoryIssue)

**Comment by sudojan** [comment]
This is probably fixed in MR #137

**Comment by Jean1995** [comment]
This has been fixed with the recent MR #137 since we are now using smart pointers instead of raw pointers.
